<a href="https://colab.research.google.com/github/Brotherswords/MeSH-Deep-Learning-Multi-Label-Classification-Model/blob/main/DeepLearning_Quiz5_Lavan_Vivekanandasarma.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Quiz 5 (Take Home) Assignment

## Installing Neccesary Libraries


In [ ]:
!pip install tensorflow
!pip install keras
!pip install nltk

In [ ]:
from google.colab import drive
import os
import zipfile
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Bidirectional
from sklearn.preprocessing import MultiLabelBinarizer
from tensorflow.keras.models import Sequential
import numpy as np
import pickle


drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Check for the file I need
directory_path = '/content/drive/My Drive/Projects Things Useful/University/CU Denver 2023-2024 Sem 1/Deep Learning/Quiz_5_Data'
pickle_path = directory_path + "/Pickle_Files"
# Check if the directory exists
if os.path.exists(directory_path):
    # List all files and directories in the specified path
    files = os.listdir(directory_path)
    print("Files and directories in '", directory_path, "' :")
    for i in files:
      print(i)
else:
    print("The directory does not exist")

Files and directories in ' /content/drive/My Drive/Projects Things Useful/University/CU Denver 2023-2024 Sem 1/Deep Learning/Quiz_5_Data ' :
training-set-100000.json
test-set-20000-rev2.json
judge-set-10000-unannotated.json
Pickle_Files
BioWordVec_PubMed_MIMICIII_d200.vec.bin


In [ ]:
import json
file_name= "/training-set-100000.json"
training_path = f'{directory_path}/{file_name}'
data = []
with open(training_path, 'r') as file:
    data = json.load(file)
    if data:
      print("Training Data Loaded")

Training Data Loaded


## Step 1. Data Pre-Processing

In [ ]:
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer, WordNetLemmatizer

# Download stopwords and wordnet data
nltk.download('stopwords')
nltk.download('wordnet')

# Initialize stemmer and lemmatizer
stemmer = PorterStemmer()
lemmatizer = WordNetLemmatizer()

# Load English stopwords
stop_words = set(stopwords.words('english'))

def clean_text(text):
    # Remove HTML tags using regex
    text = re.sub(r'<.*?>', '', text)

    # Remove special characters and digits
    text = re.sub(r'[^a-zA-Z\s]', '', text)

    # Lowercase all texts
    text = text.lower()

    # Tokenize text
    words = text.split()

    # Remove stopwords and apply stemming or lemmatization
    words = [lemmatizer.lemmatize(word) for word in words if word not in stop_words]
    # Uncomment the next line to use stemming instead of lemmatization
    # words = [stemmer.stem(word) for word in words if word not in stop_words]

    # Join the words back into a single string
    text = ' '.join(words)

    return text

# cleaned_texts = [clean_text(text) for text in texts]

# # Now `cleaned_texts` contains the preprocessed article texts


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


In [ ]:
# Assuming `data` is already loaded with the articles data in the above format
# Combine title and abstract for each article and gather labels
texts = [clean_text(d['title'] + ' ' + d['abstractText']) for d in data['articles']]
labels = [d['meshMajor'] for d in data['articles']]

# Initialize the tokenizer with a specific vocabulary size, and a filter for punctuation
vocabulary_size = 15000  # for example, you might choose the top 20,000 words
tokenizer = Tokenizer(num_words=vocabulary_size, lower=True, oov_token='UNK', filters='!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n')

tokenizer.fit_on_texts(texts)

# Convert texts to sequences of integer indices
sequences = tokenizer.texts_to_sequences(texts)
word_index = tokenizer.word_index

# Determine a suitable maximum sequence length based on the data distribution
# For instance, you can take the 90th percentile as the max length
max_seq_length = int(np.percentile([len(seq) for seq in sequences], 75))

# Pad the sequences so that they all have the same length
data_padded = pad_sequences(sequences, maxlen=max_seq_length, padding='post')

# Initialize MultiLabelBinarizer to convert the labels to a binary matrix
mlb = MultiLabelBinarizer()
label_data = mlb.fit_transform(labels)

# This absolutely devours material.
# # Save the tokenizer, MultiLabelBinarizer, max_seq_length for later use
# with open(pickle_path + '/preprocessing_objects.pickle', 'wb') as handle:
#     pickle.dump({'tokenizer': tokenizer, 'mlb': mlb, 'max_seq_length': max_seq_length,
#                  'data_padded': data_padded, 'label_data': label_data}, handle, protocol=pickle.HIGHEST_PROTOCOL)

# Now `data_padded` contains the padded sequences and `label_data` contains the multi-hot encoded labels
# `word_index` contains the word index, and `max_seq_length` is the length up to which sequences will be padded
print("Preprocessing and saving completed.")


Preprocessing and saving completed.


In [ ]:
# # Load all preprocessing objects at once
# with open(pickle_path + '/preprocessing_objects.pickle', 'rb') as handle:
#     preprocessing_objects = pickle.load(handle)

# tokenizer = preprocessing_objects['tokenizer']
# mlb = preprocessing_objects['mlb']
# max_seq_length = preprocessing_objects['max_seq_length']
# # data_padded = preprocessing_objects['data_padded']
# label_data = preprocessing_objects['label_data']
# print("Loaded")

Loaded


In [ ]:
n = 10
for i, (word, index) in enumerate(tokenizer.word_index.items()):
    print(f"{i + 1}. {word}: {index}")
    if i >= n - 1:  # since index starts at 0, we use n - 1
        break

1. UNK: 1
2. patient: 2
3. cell: 3
4. study: 4
5. result: 5
6. protein: 6
7. p: 7
8. group: 8
9. effect: 9
10. level: 10


## Step 2. Creating the Embedded Layer 🐪

### RUN THIS ONCE, AFTERWARDS JUST USE PICKLE FILE VERSION
(IT TAKES FOREVER TO LOAD)

In [ ]:
from gensim.models import KeyedVectors

embedding_path = directory_path + "/BioWordVec_PubMed_MIMICIII_d200.vec.bin"
embeddings = KeyedVectors.load_word2vec_format(embedding_path, binary=True)



# Get the word_index from your tokenizer
word_index = tokenizer.word_index

# Initialize the embedding matrix
embedding_matrix = np.zeros((len(word_index) + 1, 200))  # Add 1 for padding token

# Populate the embedding matrix
for word, i in word_index.items():
    # Check if the word is in the model
    if word in embeddings.key_to_index:
        # Get the embedding vector for the word
        embedding_vector = embeddings[word]
        # If an embedding was found, add it to the matrix
        if embedding_vector is not None:
            embedding_matrix[i] = embedding_vector

with open(pickle_path + "/embedding_matrix_15000", 'wb') as handle:
    pickle.dump(embedding_matrix, handle, protocol=pickle.HIGHEST_PROTOCOL)

print("Embedding matrix saved.")


Embedding matrix saved.


In [ ]:
# Create the embedding layer with pre-trained weights

loaded_embedding_matrix = []
with open(pickle_path + "/embedding_matrix_15000", 'rb') as handle:
    loaded_embedding_matrix = pickle.load(handle)

embedding_layer = Embedding(input_dim=len(word_index) + 1,
                            output_dim=200,
                            weights=[loaded_embedding_matrix],
                            input_length=max_seq_length,  # As determined earlier
                            trainable=False)

## Step 3. Create LSTM layers

In [ ]:
# Number of unique labels in your multi-label classification task
num_labels = len(mlb.classes_)  # Replace with actual number of labels

model = Sequential()
# Add the pre-loaded embedding layer
model.add(embedding_layer)  # embedding_layer is the one you created using loaded_embedding_matrix
# Add an LSTM layer
model.add(Bidirectional(LSTM(units=128, return_sequences=True)))
model.add(Dropout(0.1))
model.add(Bidirectional(LSTM(units=64)))
# model.add(Dropout(0.2))
# Add a dense layer
model.add(Dense(units=128, activation='relu'))
#model.add(Dropout(0.2))
# Output layer with a sigmoid activation for multi-label classification
model.add(Dense(units=num_labels, activation='sigmoid'))
# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
model.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 155, 200)          54306400  
                                                                 
 bidirectional (Bidirectiona  (None, 155, 256)         336896    
 l)                                                              
                                                                 
 dropout (Dropout)           (None, 155, 256)          0         
                                                                 
 bidirectional_1 (Bidirectio  (None, 128)              164352    
 nal)                                                            
                                                                 
 dense (Dense)               (None, 128)               16512     
                                                                 
 dense_1 (Dense)             (None, 22373)             2

## Step 4. Train the model and pray that Google Colab Pro is Enough!

In [ ]:
# Load the training data
X_train = data_padded  # Your padded sequence data
Y_train = label_data   # Your multi-hot encoded MeSH terms

In [ ]:
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.metrics import Recall

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=[Recall()])


# Set training parameters
batch_size = 32
epochs = 25
validation_split = 0.1  # Percentage of data to use as validation

# Define an EarlyStopping callback to prevent overfitting
early_stopping = EarlyStopping(monitor='val_loss', patience=3, verbose=1)

# Train the model
history = model.fit(X_train, Y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    validation_split=validation_split,
                    callbacks=[early_stopping],
                    verbose=1)

print("Training completed.")

Epoch 1/25
2813/2813 [==============================] - 553s 194ms/step - loss: 0.0068 - recall: 0.0480 - val_loss: 0.0037 - val_recall: 0.0475
Epoch 2/25
2813/2813 [==============================] - 548s 195ms/step - loss: 0.0036 - recall: 0.0492 - val_loss: 0.0036 - val_recall: 0.0576
Epoch 3/25
2813/2813 [==============================] - 547s 194ms/step - loss: 0.0034 - recall: 0.0980 - val_loss: 0.0032 - val_recall: 0.1014
Epoch 4/25
2813/2813 [==============================] - 549s 195ms/step - loss: 0.0030 - recall: 0.1249 - val_loss: 0.0029 - val_recall: 0.1302
Epoch 5/25
2813/2813 [==============================] - 547s 195ms/step - loss: 0.0028 - recall: 0.1349 - val_loss: 0.0027 - val_recall: 0.1434
Epoch 6/25
2813/2813 [==============================] - 548s 195ms/step - loss: 0.0026 - recall: 0.1453 - val_loss: 0.0026 - val_recall: 0.1520
Epoch 7/25
2813/2813 [==============================] - 548s 195ms/step - loss: 0.0025 - recall: 0.1521 - val_loss: 0.0025 - val_recall:

In [ ]:
model_save_path = pickle_path + '/saved_model_25_Epochs_Recall.h5'  # Replace with your desired path


In [ ]:
# Save the model
# V1 of model
# model_save_path = pickle_path + '/saved_model.h5'  # Replace with your desired path
model.save(model_save_path)
print(f"Model saved at {model_save_path}")


Model saved at /content/drive/My Drive/Projects Things Useful/University/CU Denver 2023-2024 Sem 1/Deep Learning/Quiz_5_Data/Pickle_Files/saved_model_25_Epochs_Recall.h5


## Step 5. Call the saved model and run it. Give it a test run to see we're getting real results.

In [ ]:
from tensorflow.keras.models import load_model

# Load the model
loaded_model = load_model(model_save_path)
print("Model loaded successfully.")

Model loaded successfully.


In [ ]:
first_sample = X_train[1:2]  # Select the first sample
predicted_result = loaded_model.predict(first_sample)  # Predict using the loaded model

# Print the predicted result
print("Predicted result:", predicted_result)

# Print the actual result
actual_result = Y_train[1]
print("Actual result:", actual_result)



1/1 [==============================] - 1s 1s/step
Predicted result: [[6.98327653e-07 5.76985667e-06 2.49171399e-06 ... 1.08561708e-05
  1.24572925e-05 1.07505741e-06]]
Actual result: [0 0 0 ... 0 0 0]


In [ ]:
# Use inverse_transform to convert binary vectors back to labels
import numpy as np

# Convert predicted result to binary and then to labels
predicted_labels_conformed = mlb.inverse_transform(np.round(predicted_result))

# Ensure actual_result is a 2D array for inverse_transform
actual_labels_conformed = mlb.inverse_transform(np.array([actual_result]))

print("Predicted MeSH terms:", predicted_labels_conformed[0])
print("Actual MeSH terms:", actual_labels_conformed[0])



Predicted MeSH terms: ('Adult', 'Aged', 'Carcinoma, Squamous Cell', 'Case-Control Studies', 'Female', 'Genetic Predisposition to Disease', 'Genotype', 'Humans', 'Male', 'Middle Aged', 'Polymorphism, Genetic', 'Uterine Cervical Neoplasms')
Actual MeSH terms: ('Adult', 'Aged', 'Aged, 80 and over', 'Carcinoma, Squamous Cell', 'Case-Control Studies', 'Female', 'Gene Frequency', 'Genetic Predisposition to Disease', 'Genotype', 'Germany', 'Greece', 'Humans', 'Interleukin-8', 'Male', 'Middle Aged', 'Mouth Neoplasms', 'Polymorphism, Genetic', 'Reverse Transcriptase Polymerase Chain Reaction', 'Risk')


## Step 4. Seeing how well we did on the test set 🙏

In [ ]:
# Load the test set
import json
file_name= "test-set-20000-rev2.json"
training_path = f'{directory_path}/{file_name}'
data = []
with open(training_path, 'r') as file:
    data = json.load(file)
    if data:
      print("Test Data Loaded")

texts = [clean_text(d['title'] + ' ' + d['abstractText']) for d in data['documents']]
labels = [d['meshMajor'] for d in data['documents']]
sequences = tokenizer.texts_to_sequences(texts)

# Pad the sequences so that they all have the same length
X_Test = data_padded_test = pad_sequences(sequences, maxlen=max_seq_length, padding='post')
Y_Test = label_data_test = mlb.transform(labels)

Test Data Loaded


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import numpy as np

# Make predictions on the test set
test_predictions = loaded_model.predict(X_Test)
binary_predictions = np.round(test_predictions)

625/625 [==============================] - 45s 70ms/step


In [ ]:
print(len(Y_Test[0]))
print(len(X_Test[0]))

print(Y_Test[0])
print(X_Test[0])

22373
155
[0 0 0 ... 0 0 0]
[2058 3561 3506 1692 3315   16    1 1301   11 1271  111 2058 3561 3506
 1692 3315  155 1188  416  983   16    1 1301  103 2863  607   42 1301
   44   62   11 3506 1247  121 2263   61    1  212 2058 3561 1692 3315
  416   81 1301   33  761  416 2263  212 1353  643 2263  106   86  985
  212   62 1188   81 1301  510 1365  111   20    5 1423 1379   10 1301
   20  766 1081  943 1301    6   62  416 2263    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0]


In [ ]:
# Calculate the metrics
accuracy = accuracy_score(Y_Test, binary_predictions)
micro_precision = precision_score(Y_Test, binary_predictions, average='micro')
micro_recall = recall_score(Y_Test, binary_predictions, average='micro')
micro_f1 = f1_score(Y_Test, binary_predictions, average='micro')

In [ ]:
print("Accuracy on Test Data:", accuracy)
print("Micro Precision on Test Data:", micro_precision)
print("Micro Recall on Test Data:", micro_recall)
print("Micro F1 Score on Test Data:", micro_f1)

Accuracy on Test Data: 0.0
Micro Precision on Test Data: 0.6947941675938397
Micro Recall on Test Data: 0.2274112492381292
Micro F1 Score on Test Data: 0.3426655422577515


# Model Report

After evaluating my model on the test data for the multi-label classification task, the results suggest that it's moderately effective, but there's significant room for improvement. Notably, the model's accuracy is surprisingly low at 0.0, indicating it's not predicting all the labels for any single sample correctly. This is a known challenge in multi-label classification, where accuracy demands a perfect match of all labels.

The micro precision, at 69.48%, shows that a majority of the predicted labels are correct. However, the micro recall is only 22.74%, implying that the model is missing many relevant labels. This is further reflected in the moderate micro F1 score of 34.27%, which balances precision and recall.

The discrepancy between precision and recall suggests a conservative model that makes fewer errors in its predictions but overlooks a significant number of correct labels. This could stem from various factors, such as the model's complexity, potential class imbalance in the dataset, or the need for more extensive training. To enhance the model's effectiveness, fine-tuning, and a more in-depth analysis are crucial next steps.



# Step 6. Making Predictions for Judge and Writing to JSON

In [ ]:
{"documents": [
{"pmid": 22511223, "labels":["label1","label2",...,"labelN"]},
{"pmid":22511224, "labels":["label1", "label2",...,"labelM"]},
...
...
.
{"pmid":22511225,"labels":["label1", "label2",...,"labelK"] }
]}

In [16]:
# Load the test set
import json
file_name= "judge-set-10000-unannotated.json"
training_path = f'{directory_path}/{file_name}'
data = []
with open(training_path, 'r') as file:
    data = json.load(file)
    if data:
      print("Judge Data Loaded")

texts = [clean_text(d['title'] + ' ' + d['abstractText']) for d in data['documents']]
labels = [d['pmid'] for d in data['documents']]
sequences = tokenizer.texts_to_sequences(texts)

# Pad the sequences so that they all have the same length
X_Judge = data_padded_test = pad_sequences(sequences, maxlen=max_seq_length, padding='post')

Judge Data Loaded


In [17]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import numpy as np

# Make predictions on the judge set
judge_predictions = loaded_model.predict(X_Judge)
binary_predictions = np.round(judge_predictions)
predicted_labels_conformed = mlb.inverse_transform((binary_predictions))

313/313 [==============================] - 22s 70ms/step


In [18]:
import json

# Prepare data for JSON export
json_output = {"documents": []}
for pmid, labels in zip(labels, predicted_labels_conformed):
    json_output["documents"].append({"pmid": pmid, "labels": list(labels)})

# Write the data to a JSON file
output_path = f'{directory_path}/predicted_labels.json'
with open(output_path, 'w') as file:
    json.dump(json_output, file, indent=4)

print(f"Predictions saved to {output_path}")


Predictions saved to /content/drive/My Drive/Projects Things Useful/University/CU Denver 2023-2024 Sem 1/Deep Learning/Quiz_5_Data/predicted_labels.json
